pkg/streamingpromql/functions/math.go

```diff
@@ -63,3 +63,23 @@ var UnaryNegation InstantVectorSeriesFunction = func(seriesData types.InstantVec
 
 	return seriesData, nil
 }
+
+var Clamp InstantVectorSeriesFunction = func(seriesData types.InstantVectorSeriesData, scalarArgsData []types.ScalarData, memoryConsumptionTracker *limiting.MemoryConsumptionTracker) (types.InstantVectorSeriesData, error) {
+	outputIdx := 0
+	for step, data := range seriesData.Floats {
+		minVal := scalarArgsData[0].Samples[step].F
+		maxVal := scalarArgsData[1].Samples[step].F
+		if maxVal < minVal {
+			// Drop this point as there is no valid answer
+			continue
+		}
+		// We reuse the existing FPoint slice in place
+		seriesData.Floats[outputIdx].F = math.Max(minVal, math.Min(maxVal, data.F))
```

Comment:
Don't we also need to set `seriesData.Floats[outputIdx].T` here? (this might be a missing test case)

Reply:
We shouldn't need to set it as we're modifying the point in place.

This should be tested by the promql range query tests. The points will not align with the timestamp if it is wrong, and if it differs to prometheus the comparison of results should catch it.

If you have an idea on how to better test it let me know.

Reply:
If `maxVal < minVal` was true for some previous steps, then we won't be modifying the point in place, we'll be shifting this point to an earlier index in `seriesData.Floats`.

Reply:
Actually, we do need this because it's T from the old point. Good catch!